### PR TITLE
Integrate `total_batch_idx` with progress tracking

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

See title

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/fit_loop.py [hunk 1]
```diff
@@ -63,12 +63,12 @@ def global_step(self, value: int) -> None:
...
     def total_batch_idx(self) -> int:
-        """Returns the total number of batches already run (across all epochs)"""
+        """Returns the current batch index (across epochs)"""
         return self.epoch_loop.total_batch_idx
 
     @property
     def batch_idx(self) -> int:
-        """Returns the number of batches already run within this epoch"""
+        """Returns the current batch index (within this epoch)"""
         return self.epoch_loop.batch_idx
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:72]: Should we remove/deprecate these properties in the fit loop?

They are not exposed in the trainer. Do we expect users to check them by accessing `trainer.fit_loop.*`?

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 2]
```diff
@@ -180,14 +185,9 @@ def on_advance_end(self):
...
         self.update_lr_schedulers("step", update_plateau_schedulers=True)
 
-        self.total_batch_idx += 1
-
         # progress global step according to grads progress
         self._increment_accumulated_grad_global_step()
 
-        if self.done:
-            raise StopIteration
-
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:190]: @awaelchli Do you know why we had this?

This is `on_advance_end` so the next thing we do is check the while loop condition

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 3]
```diff
@@ -57,6 +55,13 @@ def __init__(self, min_steps: int, max_steps: int):
...
 
+    @property
+    def total_batch_idx(self) -> int:
+        """Returns the current batch index (across epochs)"""
+        # use `ready` instead of `completed` in case this is accessed after `completed` has been increased
+        # but before the next `ready` increase
+        return self.batch_progress.total.ready - 1
+
     @property
     def batch_idx(self) -> int:
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:66]: Should we change these to protected?

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:190]: there must have been other code below these lines that since got deleted, and there was a specific reason (tests) that would need to avoid running these lines before the next done() check.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 4]
```diff
@@ -180,14 +185,9 @@ def on_advance_end(self):
         # update plateau LR scheduler after metrics are logged
         self.update_lr_schedulers("step", update_plateau_schedulers=True)
 
-        self.total_batch_idx += 1
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:183]: careful removing this, the call to `_increment_accumulated_grad_global_step` requires the updated value here.

File: pytorch_lightning/loops/fit_loop.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/loops/fit_loop.py:72]: yes, they can be removed or deprecated. 
They are simply there to simplify the previous refactors and to avoid long chains of calls through `trainer.fit_loop.some_loop.another_loop.batch_idx`. And also to make them read only when possible.

> With loop customization, the references could not exist anymore

not necessarily as we may want to introduce intermediate base classes that separate Lightning built-in loops from more versatile loops with hooks and interfaces that are more specific than the abstract Loop base class contains.

but cleaner if we can remove them :)

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:183]: @carmocca

File: tests/tuner/test_lr_finder.py [hunk 5]
```diff
@@ -184,7 +184,7 @@ def __init__(self):
 
     assert lrfinder.suggestion() != 1e-3
     assert len(lrfinder.results["lr"]) == 100
-    assert lrfinder._total_batch_idx == 200
+    assert lrfinder._total_batch_idx == 199
```
Comment (tchaton) [inline tests/tuner/test_lr_finder.py:187]: Why do we get a diff there ? Is the test deterministic ?

File: pytorch_lightning/loops/fit_loop.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/loops/fit_loop.py:72]: 👍 
I'll leave them in this PR but we should revisit all in that file eventually

File: tests/tuner/test_lr_finder.py (hunk 5 above)
Comment (carmocca) [reply tests/tuner/test_lr_finder.py:187]: > Is the test deterministic?

It is.

Seems to me like the previous `total_batch_idx` was wrong, this is because the name indicates an index but the previous value here (`200`) indicates a count.

So we could keep the old count behaviour (`200`) with a misleading variable name (`*_idx`)
or the old variable name (`*_idx`) with the fixed index behaviour (`199`).

For the LR finder in particular, seems like this attribute is just used for debugging, so should be fine to change it.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:179]: Fixed it by using `current.ready` instead of `total_batch_idx`.

Not sure why that was the case. Especially since that's different to the check done in `_should_accumulate` which does use `current.ready`.

Long term, I have opened #8856 to address these inconsistencies (WIP)

Thnaks for the eagle eye @awaelchli !

**Review by awaelchli — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]

**Review by Borda — APPROVED** [review]
